### DISCUSSION TITLE
Method loadUserByIdentifier behaves differently in UserProviderInterface and UserLoaderInterface

### DISCUSSION THREAD

**Discussion by ChMat** [symfony/symfony#52582] [Q&A]
Hi,

I’m refreshing the code base for our UserRepository and noticed a difference in the signature for methods `loadUserByIdentifier()` in the following interfaces:

- `Symfony\Component\Security\Core\User\UserProviderInterface::loadUserByIdentifier` [see file in repo](https://github.com/symfony/security-core/blob/6.4/User/UserProviderInterface.php)
- `Symfony\Bridge\Doctrine\Security\User\UserLoaderInterface::loadUserByIdentifier` [see file in repo](https://github.com/symfony/doctrine-bridge/blob/6.4/Security/User/UserLoaderInterface.php)

In `UserLoaderInterface`, it should return `null` when the user is not found.

```php
    /**
     * Loads the user for the given user identifier (e.g. username or email).
     *
     * This method must return null if the user is not found.
     */
    public function loadUserByIdentifier(string $identifier): ?UserInterface;
```

But in `UserProviderInterface`, it should throw a `UserNotFoundException`

```php
    /**
     * Loads the user for the given user identifier (e.g. username or email).
     *
     * This method must throw UserNotFoundException if the user is not found.
     *
     * @throws UserNotFoundException
     */
    public function loadUserByIdentifier(string $identifier): UserInterface;
```

At the moment, our `UserRepository` implements both interfaces (using Symfony 4.4). I’m migrating our library to Symfony 6 and I’m wondering if I should decouple the UserProvider and UserLoader rather than have the UserRepository do both jobs.

Do you think the difference is intended or is it an issue that needs a fix?

Maybe implementing `UserProviderInterface` is good enough?

Thanks for your opinions!

**Comment by Seb33300** [comment]
I don't think that you need to implement `UserProviderInterface`.

If your users come from the database, the `UserLoaderInterface` should be enough:
https://symfony.com/doc/current/security/user_providers.html#using-a-custom-query-to-load-the-user

Here is the documentation about the user provider, you need to check in your `security.yaml` if you defined it.
https://symfony.com/doc/current/security/user_providers.html#creating-a-custom-user-provider
